docs/wasm_abi_quirks.md

```diff
@@ -0,0 +1,241 @@
+%Big = type { i8, [1 x i8], i16, [2 x i16], i64 }
+define dso_local void @big(%Big %0) unnamed_addr #0 { ... }
+```
+
+And it gets invoked as `wasm.big(a, 0, b, 0, 0, c)`. Even though there are four bytes of padding for the second padding segment, it's treated as `i16`-padding, which means only two fields are needed. The padding type appears to just be the alignment of the preceding field.
+
+### Nested structs
+
+The "contains two scalar fields" rule is only applied at the top level when an aggregate is passed as an argument. If a struct with additional fields contains a struct with two scalar fields, the padding of that internal struct does become relevant again. Effectively, the actual topology of the struct is mostly irrelevant for when it is being passed over FFI, just the transitive list of fields, and any alignment/size constraints.
+
+
+
+## Unions in parameters
+
+Unions are passed as `size / align` parameters, each of size `align` for a union with size `size` and alignment `align` (This is calculated by taking the max size and max align of the two fields).
+
+For example, with the following union from Diplomat:
+
+```rust
+#[repr(C)]
+union DiplomatResultValue<T, E> {
+    ok: ManuallyDrop<T>,
+    err: ManuallyDrop<E>,
+}
+
+/// A [`Result`]-like type that can be passed across the FFI boundary
+/// as a value. Used internally to return [`Result`]s and [`Option`]s
+/// from functions.
+#[repr(C)]
+pub struct DiplomatResult<T, E> {
+    value: DiplomatResultValue<T, E>,
+    pub is_ok: bool,
+}
+
+/// A type to represent Option<T> over FFI.
+///
+/// Used internally to handle `Option<T>` arguments and return types, and needs to be
+/// used explicitly for optional struct fields.
+pub type DiplomatOption<T> = DiplomatResult<T, ()>;
+```
+
+`DiplomatOption<MyStruct>` (size 8, alignment 4) will be passed as two `u32` parameters, followed by an `u8` parameter for the `bool`, and three more padding `u8` parameters.
+
+
+Wasm is little-endian. For the following code:
+
+```rust
+#[repr(C)]
+#[derive(Debug)]
+pub struct Inner {
+    x: u8,
+    y: u16,
+    z: u32,
+}
+#[no_mangle]
+pub extern "C" fn opt(x: DiplomatOption<Inner>) {
+    let val = unsafe { x.value.ok };
+    log(&format!("{val:#x?}"));
+}
+```
+
+The call `wasm.opt(0x12345678, 0x9ABCDEF0, 1, 0, 0, 0)` (with the two u32 values corresponding to the `DiplomatResultValue` union) produces
+
+```rust
+Inner {
+    x: 0x78,
+    y: 0x1234,
+    z: 0x9abcdef0,
+}
+```
+
+with LLVM IR:
+
+```llvm
+; produced by rustc with (default) -Zwasm-c-abi=legacy
+%Inner = type { i8, [1 x i8], i16, i32 }
+%"DiplomatResult<Inner, ()>" = type { %"DiplomatResultValue<Inner, ()>", i8, [3 x i8] }
+%"DiplomatResultValue<Inner, ()>" = type { [2 x i32] }
+
+define dso_local void @opt(%"DiplomatResult<Inner, ()>" %0) unnamed_addr #0 { ... }
+```
+
+## u64
+
+While all "normal" scalar parameters in wasm are regular Number types, `u64` translates to a `BigInt`. 
```

Comment:
I think this is a browser engine thing, not a WebAssembly or Rust thing?

Reply:
It's a JS thing, not a browser thing, but I'll rephrase this.

Reply:
I moved it to the top of the file